### PR TITLE
gdk: add sanity checks for Texture::download

### LINKED ISSUE #677: Sanity check for gdk::Texture::download
see https://github.com/gtk-rs/gtk4-rs/pull/632#discussion_r734941502 for details

### PR DESCRIPTION
**Pull request by bilelmoussaoui**
Fixes #677

### CODE REVIEW

File: gdk4/src/texture.rs [hunk 1]
```diff
@@ -14,6 +15,14 @@ pub trait TextureExtManual: 'static {
 impl<O: IsA<Texture>> TextureExtManual for O {
     fn download(&self, data: &mut [u8], stride: usize) {
         unsafe {
+            assert!(
+                stride >= 4,
+                "Stride for a CAIRO_FORMAT_ARGB32 should be >= 4"
+            );
+            assert!(
+                data.len() as i32 == 4 * (stride as i32) * self.as_ref().height(),
```
Comment (sdroege) [inline gdk4/src/texture.rs:23]:
```suggestion
                data.len() as i32 >= 4 * (stride as i32) * self.as_ref().height(),
```

Maybe?

File: gdk4/src/texture.rs [hunk 2]
```diff
@@ -14,6 +15,14 @@ pub trait TextureExtManual: 'static {
 impl<O: IsA<Texture>> TextureExtManual for O {
     fn download(&self, data: &mut [u8], stride: usize) {
         unsafe {
+            assert!(
+                stride >= 4,
+                "Stride for a CAIRO_FORMAT_ARGB32 should be >= 4"
```
Comment (sdroege) [inline gdk4/src/texture.rs:20]: Also

```rust
assert!(width() * 4 <= stride);
```

File: gdk4/src/texture.rs [hunk 3]
```diff
@@ -14,6 +15,18 @@ pub trait TextureExtManual: 'static {
...
+            );
+            assert!(
+                data.len() as i32 >= 4 * (stride as i32) * self.as_ref().height(),
```
Comment (sdroege) [inline gdk4/src/texture.rs:23]:
```suggestion
                data.len() as i32 >= stride as i32 * self.as_ref().height(),
```

The stride already contains the 4

File: gdk4/src/texture.rs [hunk 4]
```diff
@@ -14,6 +15,18 @@ pub trait TextureExtManual: 'static {
...
+                "The data is not big enough to download the texture"
+            );
+            assert!(
```
Comment (sdroege) [inline gdk4/src/texture.rs:26]: And you probably want to switch the order of the second and third assertion for logical reasons

File: gdk4/src/texture.rs [hunk 5]
```diff
@@ -14,6 +15,18 @@ pub trait TextureExtManual: 'static {
 impl<O: IsA<Texture>> TextureExtManual for O {
     fn download(&self, data: &mut [u8], stride: usize) {
         unsafe {
+            assert!(
+                stride >= 4,
+                "Stride for a CAIRO_FORMAT_ARGB32 should be >= 4"
```
Comment (sdroege) [inline gdk4/src/texture.rs:20]: stride could be 0 if width is 0 or not?

Comment (bilelmoussaoui) [reply gdk4/src/texture.rs:20]: gdk has checks for ensuring the width & height are >0, I suppose this shouldn't happen

**Review by sdroege — APPROVED** [review]

**Pull request merged**